pkg/kubelet/lifecycle/handlers.go

```diff
@@ -165,3 +165,74 @@ func (a *appArmorAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult {
 		Message: fmt.Sprintf("Cannot enforce AppArmor: %v", err),
 	}
 }
+
+func NewNoNewPrivsAdmitHandler(runtime kubecontainer.Runtime) PodAdmitHandler {
+	return &noNewPrivsAdmitHandler{
+		Runtime: runtime,
+	}
+}
+
+type noNewPrivsAdmitHandler struct {
+	kubecontainer.Runtime
+}
+
+func (a *noNewPrivsAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult {
+	// If the pod is already running or terminated, no need to recheck NoNewPrivs.
+	if attrs.Pod.Status.Phase != v1.PodPending {
+		return PodAdmitResult{Admit: true}
+	}
+
+	// If the containers in a pod do not require no-new-privs, admit it.
+	if !noNewPrivsRequired(attrs.Pod) {
+		return PodAdmitResult{Admit: true}
+	}
+
+	// Make sure it is either docker or rkt runtimes.
+	if a.Runtime.Type() != kubetypes.DockerContainerRuntime && a.Runtime.Type() != kubetypes.RktContainerRuntime {
+		return PodAdmitResult{
+			Admit:   false,
+			Reason:  "NoNewPrivs",
+			Message: fmt.Sprintf("Cannot enforce NoNewPrivs: %s runtime not supported", a.Runtime.Type()),
+		}
+	}
+
+	if a.Runtime.Type() != kubetypes.DockerContainerRuntime {
+		// Make sure docker api version is valid.
+		rversion, err := a.Runtime.APIVersion()
+		if err != nil {
+			return PodAdmitResult{
+				Admit:   false,
+				Reason:  "NoNewPrivs",
+				Message: fmt.Sprintf("Cannot enforce NoNewPrivs: %v", err),
+			}
+		}
+		v, err := rversion.Compare("1.23")
+		if err != nil {
+			return PodAdmitResult{
+				Admit:   false,
+				Reason:  "NoNewPrivs",
+				Message: fmt.Sprintf("Cannot enforce NoNewPrivs: %v", err),
+			}
+		}
+		// If the version is less than 1.23 it will return -1 above.
+		if v == -1 {
+			return PodAdmitResult{
+				Admit:   false,
+				Reason:  "NoNewPrivs",
+				Message: fmt.Sprintf("Cannot enforce NoNewPrivs: docker runtime API version %q must be greater than or equal to 1.23", rversion.String()),
+			}
+		}
+	}
+
+	return PodAdmitResult{Admit: true}
+}
+
+func noNewPrivsRequired(pod *v1.Pod) bool {
```

Comment:
@jessfraz Should we handle `InitContainers` here too? Looks like, yes, we should.

Reply:
Are init containers handled differently that should probably be documented somewhere so it's not overlooked unless I missed it

Reply:
As far I understand, the most of the code from this PR is handling initContainers correctly because it works only with securityContext. The existing infrastructure is aware about containers/initContainers and executes validation on their SecurityContexts.

But when we're manually inspecting PodSpec we obligated to take care about initContainers. In this particular case, I see that `noNewPrivsRequired()` will return `false` for a pod that has initContainers with allowPrivielegeEscalation=false. I'm not sure what will be the consequence of this, perhaps `allowPrivielegeEscalation=false` won't work for a pod?